packages/teko/src/Teko_BlockDiagonalInverseOp.cpp

```diff
@@ -0,0 +1,138 @@
+// @HEADER
+// *****************************************************************************
+//      Teko: A package for block and physics based preconditioning
+//
+// Copyright 2010 NTESS and the Teko contributors.
+// SPDX-License-Identifier: BSD-3-Clause
+// *****************************************************************************
+// @HEADER
+
+#include "Teko_BlockDiagonalInverseOp.hpp"
+
+#include "Teuchos_Utils.hpp"
+
+namespace Teko {
+
+using Teuchos::RCP;
+
+/** @brief This constructor builds the block diagonal inverse operator of A.
+ *
+ * @param[in] A The block operator.
+ * @param[in] invDiag  A vector containing the inverse of each of the diagonal blocks.
+ */
+BlockDiagonalInverseOp::BlockDiagonalInverseOp(BlockedLinearOp& A,
+                                               const std::vector<LinearOp>& invDiag)
+    : invDiag_(invDiag) {
+  // sanity check
+  int blocks = blockRowCount(A);
+  TEUCHOS_ASSERT(blocks > 0);
+  TEUCHOS_ASSERT(blocks == blockColCount(A));
+  TEUCHOS_ASSERT(blocks == (int)invDiag_.size());
+
+  // create the range and product space
+  ///////////////////////////////////////////////////
+
+  // just flip flop them!
+  productRange_  = A->productDomain();
+  productDomain_ = A->productRange();
+}
+
+void BlockDiagonalInverseOp::implicitApply(const BlockedMultiVector& src, BlockedMultiVector& dst,
+                                           const double alpha, const double beta) const {
+  // call the no tranpose version
+  implicitApply(Thyra::NOTRANS, src, dst, alpha, beta);
+}
+
+/** @brief Perform a matrix vector multiply with this operator.
+ *
+ * The <code>apply</code> function takes one vector as input
+ * and applies the inverse \f$ LDU \f$ decomposition. The result
+ * is returned in \f$y\f$. If this operator is reprsented as \f$M\f$ then
```

Comment:
reprsented -> represented

I am a bit confused about the "ignoring conjugation!" part. Isn't the method using `M_trans`?

I think the docstrings should only be in the .hpp.

Reply:
I'll be honest -- most of this is copied over from https://github.com/trilinos/Trilinos/blob/26e307fb0fba3800525aef0870cea2a6dc2fd625/packages/teko/src/Teko_BlockUpperTriInverseOp.hpp#L52! 😅